File: services/notification-service/src/main/java/com/visor/school/notification/controller/NotificationController.java
package com.visor.school.notification.controller;

import com.visor.school.common.api.ApiResponse;
import com.visor.school.notification.model.Notification;
import com.visor.school.notification.service.NotificationService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Notification controller
 * Accessible by all authenticated users for their own notifications
 */
@RestController
@RequestMapping("/api/v1/notifications")
public class NotificationController {

    private final NotificationService notificationService;

    public NotificationController(NotificationService notificationService) {
        this.notificationService = notificationService;
    }

    /**
     * Get notifications for a user
     */
    @GetMapping
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<ApiResponse<List<NotificationResponse>>> getNotifications(
        @RequestParam UUID userId,
        @RequestParam(required = false, defaultValue = "false") boolean unreadOnly
    ) {
        List<Notification> notifications;
        if (unreadOnly) {
            notifications = notificationService.getUnread(userId);
        } else {
            notifications = notificationService.getByUser(userId);
        }

        List<NotificationResponse> response = notifications.stream()
            .map(NotificationResponse::from)
            .collect(Collectors.toList());

        return ResponseEntity.ok(ApiResponse.success(response));
    }

    /**
     * Mark notification as read
     */
    @PutMapping("/{id}/read")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<ApiResponse<NotificationResponse>> markAsRead(
        @PathVariable UUID id
    ) {
        Notification notification = notificationService.markAsRead(id);
        NotificationResponse response = NotificationResponse.from(notification);

        return ResponseEntity.ok(ApiResponse.success(response));
    }

    public record NotificationResponse(
        String id,
        String userId,
        String type,
        String title,
        String message,
        Map<String, Object> metadata,
        boolean read,
        String readAt,
        String createdAt
    ) {
        public static NotificationResponse from(Notification notification) {
            return new NotificationResponse(
                notification.getId().toString(),
                notification.getUserId().toString(),
                notification.getType().name(),
                notification.getTitle(),
                notification.getMessage(),
                notification.getMetadata(),
                notification.isRead(),
                notification.getReadAt() != null ? notification.getReadAt().toString() : null,
                notification.getCreatedAt().toString()
            );
        }
    }
}

File: services/notification-service/src/main/kotlin/com/visor/school/notification/controller/NotificationController.kt
package com.visor.school.notification.controller

import com.visor.school.notification.service.NotificationService
import com.visor.school.common.api.ApiResponse
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.util.UUID

/**
 * Notification controller
 * Accessible by all authenticated users for their own notifications
 */
@RestController
@RequestMapping("/api/v1/notifications")
class NotificationController(
    private val notificationService: NotificationService
) {

    /**
     * Get notifications for a user
     */
    @GetMapping
    @PreAuthorize("isAuthenticated()")
    fun getNotifications(
        @RequestParam userId: UUID,
        @RequestParam(required = false, defaultValue = "false") unreadOnly: Boolean
    ): ResponseEntity<ApiResponse<List<NotificationResponse>>> {
        val notifications = if (unreadOnly) {
            notificationService.getUnread(userId)
        } else {
            notificationService.getByUser(userId)
        }

        val response = notifications.map { NotificationResponse.from(it) }

        return ResponseEntity.ok(ApiResponse.success(response))
    }

    /**
     * Mark notification as read
     */
    @PutMapping("/{id}/read")
    @PreAuthorize("isAuthenticated()")
    fun markAsRead(
        @PathVariable id: UUID
    ): ResponseEntity<ApiResponse<NotificationResponse>> {
        val notification = notificationService.markAsRead(id)
        val response = NotificationResponse.from(notification)

        return ResponseEntity.ok(ApiResponse.success(response))
    }
}

data class NotificationResponse(
    val id: String,
    val userId: String,
    val type: String,
    val title: String,
    val message: String,
    val metadata: Map<String, Any>?,
    val read: Boolean,
    val readAt: String?,
    val createdAt: String
) {
    companion object {
        fun from(notification: com.visor.school.notification.model.Notification): NotificationResponse {
            return NotificationResponse(
                id = notification.id.toString(),
                userId = notification.userId.toString(),
                type = notification.type.name,
                title = notification.title,
                message = notification.message,
                metadata = notification.metadata,
                read = notification.read,
                readAt = notification.readAt?.toString(),
                createdAt = notification.createdAt.toString()
            )
        }
    }
}

